cpp/src/arrow/engine/expression.cc

```diff
@@ -0,0 +1,370 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/engine/expression.h"
+#include "arrow/engine/type_traits.h"
+#include "arrow/scalar.h"
+#include "arrow/type.h"
+#include "arrow/util/checked_cast.h"
+
+namespace arrow {
+namespace engine {
+
+//
+// ExprType
+//
+
+ExprType ExprType::Scalar(std::shared_ptr<DataType> type) {
+  return ExprType(std::move(type), Shape::SCALAR);
+}
+
+ExprType ExprType::Array(std::shared_ptr<DataType> type) {
+  return ExprType(std::move(type), Shape::ARRAY);
+}
+
+ExprType ExprType::Table(std::shared_ptr<Schema> schema) {
+  return ExprType(std::move(schema), Shape::TABLE);
+}
+
+ExprType::ExprType(std::shared_ptr<Schema> schema, Shape shape)
+    : schema_(std::move(schema)), shape_(shape) {
+  DCHECK_EQ(shape, Shape::TABLE);
+}
+
+ExprType::ExprType(std::shared_ptr<DataType> type, Shape shape)
+    : data_type_(std::move(type)), shape_(shape) {
+  DCHECK_NE(shape, Shape::TABLE);
+}
+
+ExprType::ExprType(const ExprType& other) : shape_(other.shape()) {
+  switch (other.shape()) {
+    case SCALAR:
+    case ARRAY:
+      data_type_ = other.data_type();
+      break;
+    case TABLE:
+      schema_ = other.schema();
+  }
+}
+
+ExprType::ExprType(ExprType&& other) : shape_(other.shape()) {
+  switch (other.shape()) {
+    case SCALAR:
+    case ARRAY:
+      data_type_ = std::move(other.data_type());
+      break;
+    case TABLE:
+      schema_ = std::move(other.schema());
+  }
+}
+
+ExprType::~ExprType() {
+  switch (shape()) {
+    case SCALAR:
+    case ARRAY:
+      data_type_.reset();
+      break;
+    case TABLE:
+      schema_.reset();
+  }
+}
+
+bool ExprType::Equals(const ExprType& type) const {
+  if (this == &type) {
+    return true;
+  }
+
+  if (shape() != type.shape()) {
+    return false;
+  }
+
+  switch (shape()) {
+    case SCALAR:
+      return data_type()->Equals(type.data_type());
+    case ARRAY:
+      return data_type()->Equals(type.data_type());
+    case TABLE:
+      return schema()->Equals(type.schema());
+    default:
+      break;
+  }
+
+  return false;
+}
+
+Result<ExprType> ExprType::CastTo(const std::shared_ptr<DataType>& data_type) const {
+  switch (shape()) {
+    case SCALAR:
+      return ExprType::Scalar(data_type);
```

Comment:
I agree that using the word "cast" with `ExprType` is misleading because it implies an operation on data, and it's only a replacement of the type and creating a new ExprType object with the same shape

Reply:
Renamed to `ExprType::WithType, WithSchema`